Better thread slugs

mxstbr opened the pull request:
<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- hyperion (frontend)

**Related issues (delete if you don't know of any)**
Closes #2770

Implements nicer thread URLs based on the scheme `/community/channel/slug(threadTitle)~threadId`. Note that I had to use tilde `~` as the delimiter between slug and id because our thread id's include hyphens, so we can't use that, and [the only other options were dots `.` or underscores `_`](https://stackoverflow.com/questions/695438/safe-characters-for-friendly-url) and both of those look worse imo.

**TODO**

- [x] Render threads at `/:community/:channel/:id`
- [x] Try and figure out if we can do `/:community/:channel/:threadSlug-:threadId` or something like that
- [x] Redirect old route to new route
  - [x] ...from frontend JS
  - ~...from hyperion?~ the frontend redirect also works on SSR, so we're good here
- [x] Redirect URLs without the title slug to the URL with the title slug
- [x] Make sure the canonical URL includes the title slug
- ~~Fix param redirect?~~ this is already handled by the redirect from the `/thread/:id` routes to the new ones. Redirecting the param to the new routes automatically would require hyperion to be connected to the db, which it currently isn't, so it's imo not worth it.
- [x] Replace all links in `src` to `/thread/xyz` with links to `/community/channel/xyz`
  - [x] Fix "Copy link to message to clipboard"
  - [x] Fix nested community and channel slugs

mxstbr committed:
Add new thread route, redirect old route to new route in frontend

mxstbr committed:
Add support custom slugs in thread URLs

mxstbr committed:
Redirect thread URL without slug to URL with slug

mxstbr committed:
Add canonical link meta tag to threads

mxstbr commented:
@brianlovin I think this is ready apart from replacing all links to `/thread/${id}` with the new format! It handles all the following cases:

- `/thread/:id`
- `/community/channel/id`
- `/community/channel/wrong-slug~id`
- `/community/wrongchannel/slug~id` (so if the thread is moved the URL redirects automatically!)
- `/wrongcommunity/wrongchannel/wrongslug~id`
- etc.

This all works because we literally ignore everything except for the ID, and then redirect to the correct URL once loaded. 😜 That makes everything work, including moving threads between channels! Can you think of anything else we might need to handle apart from the link updates?

mxstbr committed:
Fix flow

brianlovin commented:
Really interesting solution, thanks for digging into this! It's interesting because we don't actually care about the url at all when performing the fetch, but the canonical url will be super helpful with SEO. In your solution, if I'm reading this correctly, it seems like moving channels won't affect any url at all - if someone shares an old `:communityId/:oldChannelId/:threadTitle~:threadId` url, it'll still render the right thread even if the new canonical url is `:communityId/:newChannelId/:threadTitle~:threadId`...

So this seems like good work, let's give it a go! Want to update all the `src` usages to the new system, and investigate those CI failures? 

Also: am I correct that this won't change the `?thread=` url behavior for the slider?

mxstbr commented:
> Also: am I correct that this won't change the ?thread= url behavior for the slider?

Nope, it won't change anything in the slider or inbox. (or at least shouldn't...)

> Want to update all the src usages to the new system, and investigate those CI failures?

Will finish this off tomorrow!

> if someone shares an old :communityId/:oldChannelId/:threadTitle~:threadId url, it'll still render the right thread even if the new canonical url is :communityId/:newChannelId/:threadTitle~:threadId...

It will render the thread, but then redirect to `/community/newChannel/threadTitle~threadId`, if that makes sense. I just always redirect to the canonical URL if we're at a different one to circumvent that set of problems entirely.

brianlovin commented:
> It will render the thread, but then redirect to /community/newChannel/threadTitle~threadId, if that makes sense. I just always redirect to the canonical URL if we're at a different one to circumvent that set of problems entirely.

Awesome, great work man!

mxstbr committed:
Move links over to new thread url

mxstbr commented:
This will definitely need a lot of testing on alpha before shipping!

mxstbr committed:
Merge branch 'alpha' into nicer-thread-slugs

mxstbr committed:
Maybe fix e2e tests

brianlovin commented:
Looks like maybe that failing e2e test has something to do with privacy and permissions on private threads?

mxstbr commented:
Ah yeah of course, private threads can't redirect to the proper URL since the data doesn't exist 🤦‍♂️ Will dig into private thread handling!

mxstbr committed:
Render thread view at old route if non-existant

That way the error and null states propagate properly

mxstbr committed:
Persist search for message selection in thread redirect

mxstbr commented:
Aha tests still failing because I forgot to forward `?m` params, good catch e2e tests good catch! Fixed in latest commit, let's see if I forgot another edge case or not 🤞

mxstbr committed:
Fix last e2e test

mxstbr committed:
Merge branch 'alpha' into nicer-thread-slugs

mxstbr committed:
Fix typo

mxstbr commented:
Yesss now I think I've actually covered all edge cases! :100: This should also be ready for testing on alpha

mxstbr commented:
@brianlovin this is ready to ship to `hyperion.alpha.spectrum.chat` tomorrow after you give it a final review! :pray:

mxstbr committed:
Merge branch 'alpha' into nicer-thread-slugs

mxstbr commented:
Actually I might be able to make `-` work as a delimiter, since UUIDs have a consistent structure so we can have a regexp that checks for two or three of them at the end to get the ID!

Going to try that, that'd be much nicer than ~

brianlovin commented:
> Going to try that, that'd be much nicer than ~

Sounds good, I'll hang tight!

mxstbr committed:
Switch to better thread slugs structure

New thread urls look like this:

/spectrum/general/the-thread-is-too-small-mnbkzlqp-jkld-utei-tqg4-jliaeti1itud

Much better than with the tilde :+1:

mxstbr commented:
Latest commit makes it work without the tilde! Now thread URLs look like this:

- `/spectrum/general/the-thread-is-too-small-66df15e9-17aa-4f7b-ab9f-0021ca881a88`
- `/figma/feedback/i-really-really-miss-the-opacity-option-for-gradients-fff23ef5-f29f-4d6e-9614-db30b23e1f99`

😍😍😍 

This is ready for a review, and then we should get it on alpha for some solid testing!

mxstbr committed:
Remove console.log

brianlovin commented:
Hm looks like the e2e tests for the thread view are failing. I like where this is going though!

brianlovin commented on src/routes.js:
```diff
@@ -305,12 +305,11 @@ class Routes extends React.Component<Props> {
+                  // - /id-123-123-123-id => id-123-123-123-id, easy start that works
+                  // - /some-custom-slug-id-123-123-123-id => id-123-123-123-id, custom slug also works
+                  // - /-id-123-123-123-id => id-123-123-123-id, empty custom slug also works
+                  path="/:communitySlug/:channelSlug/(.*?)(\w*-\w*-\w*-\w*-\w*)"
```
Fairly sure this will break for all firebase-imported threads from way back in the day.

mxstbr commented:
Shit that's such a good catch @brianlovin, I totally forgot about those! 🤦‍♂️  Ugh.

Been brewing over how to solve that for the last couple days, here are some options I came up with:

1. Store a `slug` field on every thread in the database and expose it via the GraphQL API, both as an arg for queries (`thread(slug: "")`) and on the Thread object (`Thread.slug: LowercaseString!`).
1. Use the `~` delimited version
1. Create a migration to move all old, non-UUID threads to UUID's

What are your thoughts on either of these?

brianlovin commented:
> Store a slug field on every thread in the database and expose it via the GraphQL API

This sounds pretty risky - I like that your current version doesn't *actually* care about the slug, since it just uses the uuid for the lookup. Storing a slug and using it for lookups seems more fragile and can become stale if threads are edited.

> Use the ~ delimited version 

This seems...easiest? Just slightly less aesthetic. 

> Create a migration to move all old, non-UUID threads to UUID's

This sounds very not fun. Backfilling the usersThreads records to use updated uuids sounds like a pita and prone to bugs. Although, technically, this would be the best solution for future-proofing id patterns :/

mxstbr commented:
> Create a migration to move all old, non-UUID threads to UUID's

We have 1105 threads that have Firebase-style IDs:

```JS
db.table('threads').filter(
  db.not(db.row('id').match("^\\w{8}-\\w{4}-\\w{4}-\\w{4}-\\w{12}$"))
).count()

// => 1105
```

Actually, what's going to block this is the way we store notifications. Those have the thread IDs hardcoded in the payload string and there isn't much we can do about it :confused: Ugh

mxstbr committed:
Revert "Switch to better thread slugs structure"

This reverts commit 0327023d1594cd111ad371b94b9d76244a040536.

mxstbr committed:
Merge branch 'alpha' into nicer-thread-slugs

mxstbr commented:
Reverted back to `~` version, it seems like the best option even if not quite as aesthetically pleasing..

brianlovin commented:
Damn. I mean, we could always just burn notifications to the ground and redo those :P

mxstbr committed:
Merge branch 'alpha' into nicer-thread-slugs

mxstbr commented:
Clicked to different threads from their timestamps from the slider, got this:

![screenshot 2018-11-05 at 15 23 58](https://user-images.githubusercontent.com/7525670/48003486-dd191380-e10e-11e8-9598-7d4415cbaa11.png)

First, that shouldn't work. Second, that shouldn't happen 😅

mxstbr committed:
Fix nested thread slugs

mxstbr committed:
Fix "Copy link to message" functionality

mxstbr committed:
Fix timestamp link on DM messages

mxstbr committed:
Merge branch 'alpha' into nicer-thread-slugs

mxstbr commented:
@brianlovin this has been tested on alpha and I couldn't find any more bugs, so I'm inclined to say this is ready to ship! :tada:

brianlovin commented:
Sounds good to me! I'll do a cut today

brianlovin commented:
I found a place this breaks locally, will investigate
